google-cloud-bigtable/test/test_proxy/test_proxy.rb

```diff
@@ -0,0 +1,449 @@
+
+LOGGER = Logger.new $stderr, level: Logger::DEBUG
+
+# Injected into GRPC to enable additional logging
+module ProxyLogger
+  def logger
+    LOGGER
+  end
+end
+
+# Define a gRPC module-level logger method before grpc/logconfig.rb loads.
+module GRPC
+  extend ProxyLogger
+end
+
+# bt:: aliases Google::Cloud::Bigtable::
+def bt
+  Google::Cloud::Bigtable
+end
+
+# tp:: aliases Google::Bigtable::Testproxy::
+def tp
+  Google::Bigtable::Testproxy
+end
+
+# Helper method to provide feedback for future devs trying to maintain this file.
+# Generally we can't rely on the type system to fail fast when arguments are invalid
+# so we'll add some simple assertions to help out.
+def assert condition, what
+  return if condition
+  abort "DEBUG ASSERTION FAILED: #{what}"
+end
+
+# Implements a proxy per https://github.com/googleapis/cloud-bigtable-clients-test/blob/main/docs/test_proxy.md#
+# for client conformance testing.
+class TestProxyServer < tp::CloudBigtableV2TestProxy::Service
+  # Client provides class-level get, create and remove methods for managing client
+  # connections in a thread-safe manner, per
+  # https://github.com/googleapis/cloud-bigtable-clients-test/blob/main/docs/test_proxy.md#additional-notes
+  class Client
+    @clients = {}
+    @clients_lock = Mutex.new
+
+    def self.get id
+      @clients.fetch id
+    end
+
+    def self.create req
+      id = req.client_id
+      @clients_lock.synchronize do
+        # No tests should intentionally hit this, if we do it's our bug.
+        assert(!@clients.key?(id), "Client #{id} already exists in Client#create")
+        @clients[id] = new(**req.to_h)
+      end
+    end
+
+    def self.remove id
+      @clients_lock.synchronize do
+        assert(@clients.key?(id), "Client #{id} didn't exist in Client#remove()")
+        @clients.delete id
+      end
+    end
+
+    def close
+      # no-op?
+      # There appears to be no way to tell Bigtable to close whatever underlying connections it has
+    end
+
+    def table name
+      assert(
+        name[%r{\Aprojects/[a-z]+/instances/[a-z]+/tables/[a-z]+\z}],
+        "The `name` argument to Client#table should be in the form \"projects/X/instances/Y/tables/Z\""
+      )
+
+      @bigtable.table instance_id, name.split("/").last, app_profile_id: app_profile_id
+    end
+
+    private
+
+    attr_accessor :instance_id
+    attr_accessor :app_profile_id
+
+    def initialize client_id:, data_target:, project_id:, instance_id:, app_profile_id:, per_operation_timeout:,
+                   optional_feature_config:
+
+      LOGGER.debug "Building client #{client_id}"
+
+      # Every test uses these values
+      assert project_id == "project", "Unxpected value for `project`"
+      assert instance_id == "instance", "Unexpected value for `instance`"
+
+      # Used later in Client#table
+      self.instance_id = instance_id
+      self.app_profile_id = app_profile_id
+
+      timeout = if per_operation_timeout
```

Comment:
Is the timeout not configurable for different operations?

Reply:
Wouldn't that be working around the intent of the tests? The tests specify the timeout once when the connection is created, and the Ruby client library _claims_ to support setting a default timeout via the `connect` method; if the Ruby library doesn't then honor the timeout, I think that we've exposed a bug, that should be fixed in one place so all the deadline tests start passing.